File: ClientLibrary/config/ClientConfig.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using BaseLibrary;
using BaseLibrary.communication.protocol;

namespace ClientLibrary.config {
    /// <summary>
    /// Config for client side.
    /// </summary>
    public static class ClientConfig {
        /// <summary>
        /// Protocol witch is supported by client.
        /// </summary>
        public static readonly string[] SUPPORTED_PROTOCOLS;

        /// <summary>
        /// Protocol factory for making instance protocol by given protocol string.
        /// </summary>
        public static readonly ProtocolFactory PROTOCOL_FACTORY = new ProtocolFactory();
        
        static ClientConfig() {
            int index = 0;
            KeyValuePair<string, AProtocol>[] supportedProtocols = getSupportedProcotols();
            SUPPORTED_PROTOCOLS = new string[supportedProtocols.Length];
            foreach (var pair in supportedProtocols) {
                SUPPORTED_PROTOCOLS[index++] = pair.Key;
                PROTOCOL_FACTORY.RegisterProtocol(pair.Key, pair.Value);
            }
        }

        private static KeyValuePair<string, AProtocol>[] getSupportedProcotols() {
            List<KeyValuePair<string, AProtocol>> supportedProtocols = new List<KeyValuePair<string, AProtocol>>();
            Type [] protocolTypes = (Type[]) Assembly.GetAssembly(typeof(AProtocol))
                .GetTypes()
                .Where(t =>
                    t.IsClass &&
                    typeof(AProtocol).IsAssignableFrom(t) &&
                    t.GetCustomAttribute(typeof(ProtocolDescription)) != null)
                .ToArray();
            foreach (Type protocolType in protocolTypes) {
                AProtocol protocol = (AProtocol) Activator.CreateInstance(protocolType);
                string name =
                    ((ProtocolDescription) protocolType.GetCustomAttribute(typeof(ProtocolDescription))).NAME;
                KeyValuePair<string, AProtocol> keyValuePair = new KeyValuePair<string, AProtocol>(name, protocol);
                supportedProtocols.Add(keyValuePair);
            }
            return supportedProtocols.ToArray();
        }
    }
}

File: Java-Client/RobotBattlefield-Client-Java/src/ClientLibrary/config/ClientConfig.java

package ClientLibrary.config;

import java.util.ArrayList;
import java.util.List;

import BaseLibrary.communication.protocol.AProtocol;
import BaseLibrary.communication.protocol.ProtocolDescription;
import BaseLibrary.communication.protocol.ProtocolFactory;
import BaseLibrary.utils.ModUtils;

/**
 * Configuration for client side.
 */
public final class ClientConfig {

	private ClientConfig() {
	}

	/**
	 * Protocol witch is supported by client.
	 */
	public static final String[] SUPPORTED_PROTOCOLS;

	/**
	 * Protocol factory for making instance protocol by given protocol String.
	 */
	public static final ProtocolFactory PROTOCOL_FACTORY = new ProtocolFactory();

	static {
		int index = 0;
		KeyValuePair<String, AProtocol>[] supportedProtocols = getSupportedProcotols();
		SUPPORTED_PROTOCOLS = new String[supportedProtocols.length];
		for (KeyValuePair<String, AProtocol> pair : supportedProtocols) {
			SUPPORTED_PROTOCOLS[index++] = pair.Key;
			PROTOCOL_FACTORY.RegisterProtocol(pair.Key, pair.Value);
		}
	}

	@SuppressWarnings("unchecked")
	private static KeyValuePair<String, AProtocol>[] getSupportedProcotols() {
		List<KeyValuePair<String, AProtocol>> supportedProtocols = new ArrayList<KeyValuePair<String, AProtocol>>();

		List<Class<?>> classes = ModUtils.loadClassFromPackage("BaseLibrary.communication.protocol");
		for (Class<?> c : classes) {
			if (AProtocol.class.isAssignableFrom(c)) {
				ProtocolDescription description = c.getAnnotation(ProtocolDescription.class);
				if (description != null) {
					try {
						supportedProtocols.add(
								new KeyValuePair<String, AProtocol>(description.name(), (AProtocol) c.newInstance()));
					} catch (InstantiationException | IllegalAccessException e) {
						e.printStackTrace();
					}
				}
			}
		}

		return (KeyValuePair<String, AProtocol>[]) supportedProtocols.toArray(new KeyValuePair[0]);
	}

	private static class KeyValuePair<KEY, VALUE> {
		public final KEY Key;
		public final VALUE Value;

		public KeyValuePair(KEY key, VALUE value) {
			this.Key = key;
			this.Value = value;
		}
	}
}
